packages/flutter/lib/src/rendering/layer.dart

```diff
@@ -195,6 +196,91 @@ abstract class Layer extends AbstractNode with DiagnosticableTreeMixin {
     _needsAddToScene = false;
   }
 
+  Layer _highlightConflictingLayer(PhysicalModelLayer child) {
+    final ui.PictureRecorder recorder = ui.PictureRecorder();
+    final Canvas canvas = Canvas(recorder);
+    canvas.drawPath(
+      child.clipPath,
+      Paint()
+        ..color = const Color(0xFFAA0000)
+        ..style = PaintingStyle.stroke
+        ..strokeWidth = child.elevation + 10.0,
+    );
+    final PictureLayer pictureLayer = PictureLayer(child.clipPath.getBounds())
+      ..picture = recorder.endRecording()
+      ..debugCreator = child;
+    child.append(pictureLayer);
+    return pictureLayer;
+  }
+
+  List<Layer> _processConflictingPhysicalLayers(PhysicalModelLayer predecessor, PhysicalModelLayer child) {
+    FlutterError.reportError(FlutterErrorDetails(
+      exception: FlutterError('Painting order is out of order with respect to elevation.'),
+      library: 'rendering library',
+      context: 'during compositing',
+      informationCollector: (StringBuffer buffer) {
+        buffer.writeln('Attempted to composite layer:');
+        buffer.writeln(child);
+        buffer.writeln('after layer:');
+        buffer.writeln(predecessor);
+        buffer.writeln('which occupies the same area at a higher elevation.');
+      }
+    ));
+    return <Layer>[
+      _highlightConflictingLayer(predecessor),
+      _highlightConflictingLayer(child),
+    ];
+  }
+
+  /// Checks that no [PhysicalModelLayer] would paint after another
+  /// [PhysicalModelLayer] that has a higher elevation.
+  ///
+  /// Returns a list of [Layer] objects it added to the tree to highlight bad
+  /// nodes.  These layers should be removed from the tree after it has been
+  /// shipped to the engine.
+  List<Layer> debugCheckElevations() {
+    if (this is! ContainerLayer) {
+      return null;
+    }
+    final List<PhysicalModelLayer> physicalModelLayers = <PhysicalModelLayer>[];
+    final List<Layer> addedLayers = <Layer>[];
+    final ContainerLayer container = this;
+    bool _predecessorIsNotDirectAncestor(Layer predecessor, Layer child) {
+      while (child != null) {
+        if (child == predecessor) {
+          return true;
+        }
+        child = child.parent;
+      }
+      return false;
+    }
+
+    for (Layer child in container.depthFirstIterateChildren()) {
+      if (child is PhysicalModelLayer) {
```

Comment:
If you only care about the PhysicalModelLayers, then use [`Iterable.whereType()`](https://api.dartlang.org/stable/2.2.0/dart-core/Iterable/whereType.html)